### PR TITLE
Buyer order tracking

### LINKED ISSUE #91: Buyer-Order Tracking
### Description
 This backend is  to handle the checkout process, including validating and processing the order, updating inventory and updating the customer's account. 

### Functionalities

1. Make an order
   - Allow buyers to select products and add them to their chart.
   - Calculate the total amount by multiplying the product price with the quantity selected and as well add all products's prices to find the total amount.
   - Deduct the total amount from the buyer's account balance.
   - Send an email notification confirming the purchase, including details of product bought and payment made.
   - Update the order status on the application to transit.

2. Track order status
     
   - Enable buyer to track the status of their order.
   - Provide real-time updates on the order status, including if it is in transit or has arrived
   - Allow the buyer to confirm receipt of the product, marking the order status as delivered.
   - Facilitate product returns, updating the order status accordingly on the app. 

### Acceptance criteria
 
- Buyers can successfully make orders by adding products to their charts and completing the checkout process
- The total amount is calculated and deducted from the buyer's account balance
- Emails are sent to the buyers confirming their order on the application, with real-time updates on order status changes.
- Order status is updated to  show current stage of the delivery including in transit, delivered, returned.

### Definition of Done: 

- [ ]  Backend API endpoints for order processing and order tracking are implemented and functional.
- [ ]  Order processing logic  handles deductions from the buyer's account balance and updates inventory levels
- [ ]  Email notification system is integrated to send confirmations to buyers upon successful orders.
- [ ]  Real-time updates for order status tracking are implemented using various technology.
- [ ]  testing is conducted to ensure reliability and usability.
- [ ]  Feedback from buyers is gathered and incorporated to enhance the checkout and order tracking experience.

### PR DESCRIPTION
**Pull request by UwicyezaG**
### Description
 This backend is  to handle the checkout process, including validating and processing the order, updating inventory and updating the customer's account. 

### Functionalities

1. Make an order
   - Allow buyers to select products and add them to their chart.
   - Calculate the total amount by multiplying the product price with the quantity selected and as well add all products's prices to find the total amount.
   - Deduct the total amount from the buyer's account balance.
   - Send an email notification confirming the purchase, including details of product bought and payment made.
   - Update the order status on the application to transit.

2. Track order status
     
   - Enable buyer to track the status of their order.
   - Provide real-time updates on the order status, including if it is in transit or has arrived
   - Allow the buyer to confirm receipt of the product, marking the order status as delivered.
   - Facilitate product returns, updating the order status accordingly on the app. 

### Acceptance criteria
 
- Buyers can successfully make orders by adding products to their charts and completing the checkout process
- The total amount is calculated and deducted from the buyer's account balance
- Emails are sent to the buyers confirming their order on the application, with real-time updates on order status changes.
- Order status is updated to  show current stage of the delivery including in transit, delivered, returned.

### Definition of Done: 

- [ ]  Backend API endpoints for order processing and order tracking are implemented and functional.
- [ ]  Order processing logic  handles deductions from the buyer's account balance and updates inventory levels
- [ ]  Email notification system is integrated to send confirmations to buyers upon successful orders.
- [ ]  Real-time updates for order status tracking are implemented using various technology.
- [ ]  testing is conducted to ensure reliability and usability.
- [ ]  Feedback from buyers is gathered and incorporated to enhance the checkout and order tracking experience.

### CODE REVIEW

**Comment by faid-terence** [comment]
@UwicyezaG and @elijahladdie Rename the PR !
Follow best PR naming conventions

**Comment by faid-terence** [comment]
@UwicyezaG and @elijahladdie 
What's the progres ??

**Comment by faid-terence** [comment]
Too many commit messages squash them we need only one commit message !

**Comment by faid-terence** [comment]
Follow this https://github.com/atlp-rwanda/engineering-playbook/wiki/3.-Pull-Request for PR naming

**Comment by faid-terence** [comment]
@elijahladdie and @UwicyezaG 
What are the updates on this PR ?

**Review by elijahladdie — APPROVED** [review]

**Comment by faid-terence** [comment]
@UwicyezaG  and @elijahladdie 
Test coverage decreased by 2.9% can you increase it by adding more test cases !

File: src/entities/User.ts
Comment (faid-terence) [inline src/entities/User.ts:1]: @elijahladdie 
Please avoid using `any` in typescript , please define order interface

File: src/utils/index.ts
Comment (faid-terence) [inline src/utils/index.ts:1]: @UwicyezaG  and @elijahladdie 
amount should be number not any !

File: src/utils/sendOrderMail.ts
Comment (faid-terence) [inline src/utils/sendOrderMail.ts:1]: orderDate should be type `Date` not any

File: src/utils/sendOrderMailUpdated.ts
Comment (faid-terence) [inline src/utils/sendOrderMailUpdated.ts:1]: Remove any

**Review by faid-terence — CHANGES_REQUESTED** [review]
@UwicyezaG  and @elijahladdie 
Apply the feedbacks

**Review by Ndevu12 — APPROVED** [review]

File: src/utils/sendOrderMail.ts
Comment (elijahladdie) [reply src/utils/sendOrderMail.ts:1]: Feedback solved @faid-terence !

File: src/entities/User.ts [hunk 1]
```diff
@@ -98,12 +99,18 @@ export class User {
...
+  transactions!: Transaction[];
+
   @CreateDateColumn()
   createdAt!: Date;
 
   @UpdateDateColumn()
   updatedAt!: Date;
 
+  @Column({ type: 'numeric', precision: 24, scale: 2, default: 0 })
+  accountBalance!: number;
```
Comment (aimedivin) [inline src/entities/User.ts:112]: @elijahladdie , @UwicyezaG How this field will get it's value

File: src/services/orderServices/updateOrderService.ts [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+            if (!order) {
+                return sendErrorResponse(res, 404, "Order not found");
+            }
+            // Check if order can be updated
+            if (isOrderFinalStatus(order.orderStatus)) {
+                return sendErrorResponse(res, 401, `Order cannot be updated once it is ${order.orderStatus}`);
+            }
+
+            // Handle order status transitions
+            if (orderStatus !== undefined && order.orderStatus !== orderStatus) {
```
Comment (aimedivin) [inline src/services/orderServices/updateOrderService.ts:45]: @elijahladdie , @UwicyezaG , Before updating the order status, could you please check if the order has been marked as "received" first? It doesn't make sense for a buyer to update an order to "canceled" after they have already confirmed that they received it.

**Review by aimedivin — CHANGES_REQUESTED** [review]

File: src/services/orderServices/updateOrderService.ts [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+import { Product } from '../../entities/Product';
+import { User } from '../../entities/User';
+import { OrderItem } from '../../entities/OrderItem';
+import { Transaction } from '../../entities/transaction';
+import { responseError, sendErrorResponse, sendSuccessResponse } from '../../utils/response.utils';
+import sendMail from '../../utils/sendOrderMail';
+
+export const updateOrderService = async (req: Request, res: Response) => {
+    const { orderId } = req.params;
+    const { orderStatus } = req.body;
```
Comment (aimedivin) [inline src/services/orderServices/updateOrderService.ts]: @UwicyezaG , @elijahladdie Can you check if provided orderStatus is among the ones defined as possible value in "orderStatus" field

**Review by aimedivin — CHANGES_REQUESTED** [review]

**Review by aimedivin — APPROVED** [review]

**Review by MC-Knight — APPROVED** [review]

**Review by faid-terence — APPROVED** [review]
PR Appproved @UwicyezaG and @elijahladdie !

**Pull request merged**